cmd/chaos-multiversion-helper/autoconvert/generateconvert.go

```diff
@@ -0,0 +1,633 @@
+
+func generateConvert(log logr.Logger, version string, hub string) error {
+	fileSet := token.NewFileSet()
+
+	apiDirectory := "api" + "/" + version
+	sources, err := ioutil.ReadDir(apiDirectory)
+	if err != nil {
+		return errors.Wrapf(err, "read directory %s", apiDirectory)
+	}
+
+	definitionMap := map[string]*ast.GenDecl{}
+	types := newUniqueList()
+	for _, file := range sources {
+		// skip files which are not golang source code
+		if !strings.HasSuffix(file.Name(), "go") {
+			continue
+		}
+
+		filePath := apiDirectory + "/" + file.Name()
+		fileAst, err := parser.ParseFile(fileSet, filePath, nil, parser.ParseComments)
+		if err != nil {
+			return errors.Wrapf(err, "parse file %s", filePath)
+		}
+
+		// add all type definition to the definitionMap
+		ast.Inspect(fileAst, func(n ast.Node) bool {
+			node, ok := n.(*ast.GenDecl)
+			if !ok || node.Tok != token.TYPE {
+				return true
+			}
+
+			typeName := node.Specs[0].(*ast.TypeSpec).Name.Name
+			definitionMap[typeName] = node
+
+			return false
+		})
+
+		// read the comment map to decide which types need to be converted
+		cmap := ast.NewCommentMap(fileSet, fileAst, fileAst.Comments)
+		for node, commentGroups := range cmap {
+			node, ok := node.(*ast.GenDecl)
+			if !ok || node.Tok != token.TYPE {
+				continue
+			}
+
+			typeName := node.Specs[0].(*ast.TypeSpec).Name.Name
+			for _, commentGroup := range commentGroups {
+				isObjectRoot := false
+
+				for _, comment := range commentGroup.List {
+					if strings.Contains(comment.Text, "+kubebuilder:object:root=true") {
+						isObjectRoot = true
+					}
+				}
+
+				if isObjectRoot {
+					types.push(typeName)
+				}
+			}
+		}
+	}
+
+	convertFilePath := apiDirectory + "/" + "zz_generated.convert.chaosmesh.go"
+	convertFile, err := os.Create(convertFilePath)
+	if err != nil {
+		return errors.Wrapf(err, "create file %s", convertFilePath)
+	}
+	defer convertFile.Close()
+
+	impl := convertImpl{
+		definitionMap: definitionMap,
+		version:       version,
+		hub:           hub,
+		types:         types,
+		log:           log,
+	}
+
+	return impl.generate(convertFile)
+}
+
+type convertImpl struct {
+	definitionMap map[string]*ast.GenDecl
+
+	version, hub string
+
+	types *uniqueList
+
+	log logr.Logger
+}
+
+func (c *convertImpl) generate(convertFile *os.File) error {
+	// mark the original added types.
+	// the signature of these types will be different from other `Convert` function.
+	// because the `controller-runtime` requires them to convert to the `conversion.Hub`
+	hubTypes := make(map[string]struct{})
+
+	err := c.types.forEach(func(typ string) error {
+		hubTypes[typ] = struct{}{}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	convertFile.WriteString(common.Boilerplate + "\n")
+	convertFile.WriteString("package " + c.version + "\n\n")
+	convertFile.WriteString(`
+import (
+	"github.com/chaos-mesh/chaos-mesh/api/` + c.hub + `"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+)
+	`)
+
```

Comment:
I would like to add a line for each type to indicate the implemented interface, like `var _ conversation.Convertible = &typ{}`.